File: lib/leetcode/124_binary-tree-maximum-path-sum.rb
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer}

def max_path_sum(root)
  @max = -Float::INFINITY
  max_path_recursive(root)

  @max
end

def max_path_recursive(node)
  return 0 if node.nil?

  # possibilities
  # 1. self
  # 2. left + self
  # 3. right + self
  # 4 self + left + right
  left = max_path_recursive(node.left)
  right = max_path_recursive(node.right)

  max_path = [node.val, node.val + left, node.val + right].max

  @max = [max_path, @max, node.val + left + right].max

  max_path
end

File: lib/leetcode/124_binary-tree-maximum-path-sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
const maxPathSum = function(root) {
  let res = Number.MIN_SAFE_INTEGER;
  const dfs = function(node) {
    if (!node) {
      return 0;
    }

    let left = Math.max(0, dfs(node.left));
    let right = Math.max(0, dfs(node.right));

    res = Math.max(res, left + node.val + right);

    return Math.max(left, right) + node.val;
  };

  dfs(root);

  return res;
};
